src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlBatch.cs

```diff
@@ -0,0 +1,234 @@
+            : this()
+        {
+            Connection = connection;
+            Transaction = transaction;
+        }
+
+        public override int Timeout
+        {
+            get
+            {
+                CheckDisposed();
+                return _batchCommand.CommandTimeout;
+            }
+            set
+            {
+                CheckDisposed();
+                _batchCommand.CommandTimeout = value;
+            }
+        }
+
+        protected override DbBatchCommandCollection DbBatchCommands => _providerCommands != null ? _providerCommands : _providerCommands = new SqlBatchCommandCollection(Commands); // Commands call will check disposed
+
+        protected override DbConnection DbConnection
+        {
+            get
+            {
+                CheckDisposed();
+                return Connection;
+            }
+            set
+            {
+                CheckDisposed();
+                Connection = (SqlConnection)value;
+            }
+        }
+        protected override DbTransaction DbTransaction
+        {
+            get
+            {
+                CheckDisposed();
+                return Transaction;
+            }
+            set
+            {
+                CheckDisposed();
+                Transaction = (SqlTransaction)value;
+            }
+        }
+
+        public override void Cancel()
+        {
+            CheckDisposed();
+            _batchCommand.Cancel();
+        }
+
+        public override int ExecuteNonQuery()
+        {
+            CheckDisposed();
+            SetupBatchCommandExecute();
+            return _batchCommand.ExecuteNonQuery();
+        }
+
+        public override Task<int> ExecuteNonQueryAsync(CancellationToken cancellationToken = default)
+        {
+            CheckDisposed();
+            SetupBatchCommandExecute();
+            return _batchCommand.ExecuteNonQueryAsync(cancellationToken);
+        }
+
+        public override object ExecuteScalar()
+        {
+            CheckDisposed();
+            SetupBatchCommandExecute();
+            return _batchCommand.ExecuteScalar();
+        }
+
+        public override Task<object> ExecuteScalarAsync(CancellationToken cancellationToken = default)
+        {
+            CheckDisposed();
+            SetupBatchCommandExecute();
+            return _batchCommand.ExecuteScalarBatchAsync(cancellationToken);
+        }
+
+        public override void Prepare()
+        {
+            CheckDisposed();
+        }
+
+        public override Task PrepareAsync(CancellationToken cancellationToken = default)
+        {
+            CheckDisposed();
+            return Task.CompletedTask;
+        }
+
+        public override void Dispose()
+        {
+            _batchCommand?.Dispose();
+            _batchCommand = null;
+            _commands?.Clear();
+            _commands = null;
+            base.Dispose();
+        }
+
+        public List<SqlBatchCommand> Commands
```

Comment:
Both Npgsql and MySqlConnector call this `new BatchCommands` and return a custom collection type that derives from `DbBatchCommandCollection`.

https://github.com/npgsql/npgsql/blob/8d6da14034aedb88ec5ad98474a99fa38e670535/src/Npgsql/NpgsqlBatch.cs#L20
https://github.com/mysql-net/MySqlConnector/blob/12d6f818a94f4006ddaaa00ecf5592205408975c/src/MySqlConnector/MySqlBatch.cs#L96

Reply:
See https://github.com/dotnet/runtime/issues/28633#issuecomment-864630492 for rationale behind exposing custom class, rather than just `IList<T>`.

Reply:
Done.